### ISSUE TITLE
Pytree Protocol

### ISSUE THREAD

**Issue by cgarciae** [google/jax#8099]
Hey! I'd like to propose that apart from the current registration mechanism JAX allowed users to define pytrees via a Pytree Protocol. There are two reasons for doing this:

1. Simplifies the definition of custom Pytrees. Currently the best mechanism is to ensure that a class and all its subclasses are registered is to override the `__init_subclass__` method and perform the registration there, implementing a protocol is more straightforward as `__init_subclass__` is not widely known. 
2. As the concept of a Pytree becomes more wide spread within the Python ecosystem (e.g. pytorch/pytorch#65761, and `dm-tree`) a protocol could be a simple way of getting cross-library compatibility if the different implementations adopt it.

## Implementation
The idea would be to take the same mechanism implemented in `register_pytree_node_class` but with "special methods":

```python
class Pytree(Protocol):
    def __tree_flatten__(self) -> Tuple[Sequence[Any], Any]:
        ...
    
    @classmethod
    def __tree_unflatten__(cls, aux: Any, children: Sequence[Any]) -> Any:
        ...
```

Any object that defines these method would be treated as a Pytree at runtime.

## Other comments/ideas
* Pytree seems like a general concept orthogonal to numeric computing, making JAX's Pytree implementation a separate project that JAX depended on would have a positive impact on the whole Python ecosystem.
* Libraries that focus on general Pytree manipulation like [Treeo](https://github.com/cgarciae/treeo) could start to be useful independent of JAX.

**Comment by hawkinsp** [comment]
That's a good idea, especially if libraries other than JAX are going to start using pytrees.

One thing I will mention is that I don't think it makes sense to have a single global definition of what a pytree is. JAX pytrees have semantics that make sense *for JAX APIs*. But one can imagine that one might want to treat a class as a container for JAX, but treat it as an opaque object for some other API or vice versa. So I suspect `__tree_flatten__` might need to receive an object which is a description of the flattener (yet to exist) so it can decide whether it wants to be flattened by that API.

I suspect this will become clearer as part of discussions about perhaps splitting pytree into its own library.

Another fix: please flip the order of `aux` and `children` in `__tree_unflatten__`. It has always bothered me that JAX's flatten and unflatten have different orders, but I haven't wanted the disruption of changing it.

**Comment by cgarciae** [comment]
@hawkinsp updated the signature, I was tempted to do the fix but didn't know if there was a reason for the asymmetry.

**Comment by froystig** [comment]
> One thing I will mention is that I don't think it makes sense to have a single global definition of what a pytree is. JAX pytrees have semantics that make sense _for JAX APIs_. But one can imagine that one might want to treat a class as a container for JAX, but treat it as an opaque object for some other API or vice versa. So I suspect `__tree_flatten__` might need to receive an object which is a description of the flattener (yet to exist) so it can decide whether it wants to be flattened by that API.

To add to the motivation here: this protocol nicely handles custom types, but sometimes a system like JAX wants to define pytree rules (flatten/unflatten) for types whose definition it doesn't control. This includes built-ins like `tuple` and `dict`. We handle these today by maintaining an external registry that maps _any_ type to its flattening rules. This registry is essentially a hidden input to `tree_flatten` everywhere.

If we introduce this protocol but a type doesn't implement it, the flattening logic can still back off to the externally registered rules. Per @hawkinsp's point about avoiding global pytree semantics, it makes sense for every project to maintain its own such registry.

What could we pass in to `__tree_flatten__` to identify the flattener? I'm not sure; maybe some instance of a `PytreeContext` class that the pytree library defines. Whatever it is, let's have it contain a reference to a registry in order for recursive flattening to be coherent?

**Comment by cgarciae** [comment]
@froystig My initial thought was that this protocol would be used with unregistered types that implement it, but the registration mechanism would still be around because of what you say about basic python types.

If the protocol will me expanded with an additional metadata field it would be nice if additional information from the caller like `is_leaf` is passed through it to the user. As a motivation for this, I found a solution to #7919 but it didn't work if `is_leaf` was passed to `tree_flatten` because jax currently doesn't forward that information.

**Comment by froystig** [comment]
> @froystig My initial thought was that this protocol would be used with unregistered types that implement it, but the registration mechanism would still be around because of what you say about basic python types.

What I was hoping to convey is not only that a registry would still be around, but that we might consider explicitly passing it to `tree_flatten` and to the flatten/unflatten methods of the protocol in order to keep recursive flattening well-defined. JAX's registry is currently a hidden input to the pytree library, but if we fork a library for pytrees (per https://github.com/pytorch/pytorch/issues/65761#issuecomment-936507005) then there will naturally be several such registries.

**Comment by patrick-kidger** [comment]
@hawkinsp mentions sorting out the order of arguments to `tree_unflatten`. If we're willing to make a compatibility break like that, then there's a couple of other things that I think might be desirable to sort out at the same time:

**"PyDags"**
(c.f. #7919)
At present `unflatten(flatten(...))` is not the identity function.

It will create separate copies of elements that appear multiple times in the tree. There are times when this is undesirable; a notable one is when using PyTrees to represent parameterised functions -- e.g. a PyTorch-like Module system -- a la [Equinox](https://github.com/patrick-kidger/equinox), [Treex](https://github.com/cgarciae/treex) etc.

In principle I don't think this should break any existing PyTree uses: these assume referential transparency so they should be agnostic to whether they get a copy of each leaf/subtree or not.

**No auxiliary data**
This is something I've been thinking about for a while now. My current belief is that the auxiliary data output is actually unnecessary.

I think this is muddling up two separate notions: (a) partitioning a PyTree into two (or even `n`) pieces; (b) (un)flattening a PyTree. A simpler/neater API is to have *everything be a leaf*, and then invoke some `tree1, tree2 = partition(tree, partition_rule)` -- or even some `tree1, tree2, tree3 = partition(...)` etc. -- and then pass `tree1` and `tree2` separately across an API boundary. Each of `tree1`, `tree2` are PyTrees with the same structure as `tree`; each contain some subset of the leaves and have some dummy value for the leaves they don't have.

The above discussion on custom flatteners corresponds to a choice of partition rule.

**Comment by cgarciae** [comment]
> PyDags

I totally agree with @patrick-kidger that if there is going to be a refactor / formalization its a great opportunity to support PyDags.

> No auxiliary data

This would mean that in general for `Mapping`s, keys would be leaves which would be very inconvenient for a simple `tree_map` over a `dict`.

**Comment by patrick-kidger** [comment]
> This would mean that in general for `Mapping`s, keys would be leaves which would be very inconvenient for a simple `tree_map` over a `dict`.

I think this should be doable. Choosing just the values would be part of a partition rule. (I think the cleanest API for this is still up for debate, though.) Conversely, having the keys static means that it is currently impossible to do a `tree_map` over the keys of a `dict`.

**Comment by purepani** [comment]
Since scipy has been converting stuff to the array api, there have been a few instances where creating a class as a pytree would be quite helpful. In particular, any object api would likely benefit from it(Particularly the Rotation/RigidTransform classes and RBFInterpolator have been encountered so far). 

Right now, it can't quite be done yet quite so easily since importing jax within scipy to register the class as a pytree isn't desirable, but having a protocol defining a pytree(ideally just defining `__tree_flatten__` and `__tree_unflatten__` on a class) which doesn't have to be registered would allow for this. 

There is also https://github.com/metaopt/optree now which is acting as an external pytree library.
